### PR TITLE
PLANET-6373 Blocks as a theme feature POC

### PR DESCRIPTION
**Pull request by sagarsdeshmukh**
Ref. https://jira.greenpeace.org/browse/PLANET-6373

- The Guestbook block is added in Master theme as POC of Blocks as a theme feature

**File/folder structure-** 
Backend-            
image

Frontend-
image

Template-
image

**Testing**
- Deactivate the Blocks plugin
- Go to WP admin and add new page
- Add Guestbook block on page
- Publish the page and check Guestbook block on frontend
eg. https://www-dev.greenpeace.org/test-sinope/test-guestbook-block-as-a-part-of-mastertheme/

### CODE REVIEW

**Comment by planet-4** [comment]
### Test instance is ready :rocket:

:new_moon: [sinope](https://www-dev.greenpeace.org/test-sinope) | [admin](https://www-dev.greenpeace.org/test-sinope/wp-admin/) | [blocks report](https://www-dev.greenpeace.org/test-sinope/wp-admin/?page=plugin_blocks_report) | [CircleCI](https://app.circleci.com/pipelines/github/greenpeace/planet4-test-sinope) | [composer-local.json](https://github.com/greenpeace/planet4-test-sinope/blob/main/composer-local.json)

:watch: 2023.05.25 09:45:20

File: assets/src/blocks/GuestBook/GuestBookBlock.js [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+    supports: {
+      html: false, // Disable "Edit as HTMl" block option.
+      multiple: false, // Use the block just once per post.
+    },
+    edit: () => (
+      <p className="EmptyMessage">
+        {__('This block only renders in the frontend', 'planet4-blocks-backend')}
+      </p>
+    ),
+    save: frontendRendered(BLOCK_NAME),
```
Comment (GP-Dan-Tovbein) [inline assets/src/blocks/GuestBook/GuestBookBlock.js:21]: What is the main reason to use a deprecated function?

File: assets/src/blocks/frontendIndex.js [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+      if (!COMPONENTS[blockName]) {
+        return;
+      }
+
+      const BlockFrontend = COMPONENTS[blockName];
+      if (!BlockFrontend) {
+        return;
+      }
+      const attributes = JSON.parse(blockNode.dataset.attributes);
+      wp.element.render(<BlockFrontend {...attributes.attributes} />, blockNode);
```
Comment (GP-Dan-Tovbein) [inline assets/src/blocks/frontendIndex.js:21]: Correct me if I'm wrong, but I think `hydrate` instead and then change `date-render` to `data-hydrate` to all the occurrences.

File: assets/src/blocks/GuestBook/GuestBookBlock.js (hunk 1 above)
Comment (sagarsdeshmukh) [reply assets/src/blocks/GuestBook/GuestBookBlock.js:21]: Not a specific reason, I just pick the code from current [GuestBook block](https://github.com/greenpeace/planet4-plugin-gutenberg-blocks/blob/main/assets/src/blocks/GuestBook/GuestBookBlock.js#L21).

**Comment by sagarsdeshmukh** [comment]
Hi @lithrel , I rebased the PR & added a feature toggle option (with migration script to make Planet4Blocks default enabled), please have a look. also any feedback on folder structure changes will be appreciated.

**Review by lithrel — COMMENTED** [review]
Code and structure look good to me :)
Maybe wait for [this Guestbook refactoring](https://github.com/greenpeace/planet4-plugin-gutenberg-blocks/pull/1105) before concluding this PR.

I tried to see if we could use this opportunity to migrate blocks to a more [Gutenberg way](https://developer.wordpress.org/block-editor/reference-guides/block-api/block-metadata/), with a `block.json` file and editor/view scripts automatically enqueued; but this implies solving a few issues like webpack compilation so I think it's safer to migrate things first and solve that in another ticket :+1:

File: assets/src/blocks/frontendIndex.js (hunk 2 above)
Comment (sagarsdeshmukh) [reply assets/src/blocks/frontendIndex.js:18]: As discussed on DM with @GP-Dan-Tovbein , we will follow up on this change, whenever [PLANET-6917](https://jira.greenpeace.org/browse/PLANET-6917) is merged.

File: src/Blocks/BaseBlock.php [hunk 3]
```diff
@@ -120,4 +112,136 @@ public static function render_frontend(array $attributes): string
...
+        $filepath = static::get_dir_path() . 'EditorScript.js';
+        if (! file_exists($filepath)) {
+            return;
+        }
+
+        wp_enqueue_script(
+            static::get_full_block_name() . '-editor-script',
+            static::get_url_path() . 'EditorScript.js',
+            'planet4-blocks-editor-script',
+            \P4GBKS\Loader::file_ver($filepath),
```
Comment (sagarsdeshmukh) [inline src/Blocks/BaseBlock.php:161]: Hi @lithrel , here we can use a master-theme file_ver function (i.e [theme_file_var()](https://github.com/greenpeace/planet4-master-theme/blob/main/src/Loader.php#L190)), right?

File: src/Loader.php [hunk 4]
```diff
@@ -147,6 +144,19 @@ public function get_services(): array
...
+    /**
+     * Load blocks from Theme.
+     */
+    public static function add_blocks(): void
+    {
+        if (!Planet4Blocks::is_active()) {
+            return;
+        }
+
+        new MasterBlocks();
```
Comment (sagarsdeshmukh) [inline src/Loader.php:156]: The sonarcloud is complaining here -

> Either remove this useless object instantiation of class "MasterBlocks" or use it

Is it Ok, if we ignore it by adding a NOSONAR comment?
`new MasterBlocks();//NOSONAR`
cc: @lithrel

Comment (lithrel) [reply src/Loader.php:156]: Yes :) either that or making those class methods static, anything works fine :+1:

File: src/Blocks/BaseBlock.php (hunk 3 above)
Comment (lithrel) [reply src/Blocks/BaseBlock.php:161]: They don't take the same argument (one takes the full filepath, the other a path relative to the theme), so we need to adapt one or the other; but after that it's the same `filectime` principle yes

Comment (sagarsdeshmukh) [reply src/Blocks/BaseBlock.php]: I updated the PR, please have a look.

**Comment by lithrel** [comment]
@sagarsdeshmukh  Looking good :+1: 
My main issue is that I had to change the `npm install` command adding `--legacy-peer-deps` because of some React version conflict between packages. We should look into that in a different ticket, and communicate about that with the rest of the team

**Review by lithrel — APPROVED** [review]
:+1:

**Pull request merged**